### PR TITLE
lds: ASAN repro of heap-use-after-free in LDS and a single character fix

### PR DESCRIPTION
**Pull request by penguingao**
Signed-off-by: pengg <pengg@google.com>

Commit Message: Fix a heap-use-after-free in LDS
Additional Description: ListenSocketFactoryImpl stores a const std::string& to the first Listener, which is destructed upon LDS update. When reuse_port is enabled, ListenSocketFactoryImpl reads the stale reference which may be a huge garbage string depending on where std::string::size() is aligned to the garbage memory region.
Risk Level: Low
Testing: Integration with ASAN to reproduce
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features:

### CODE REVIEW

**Comment by penguingao** [comment]
cc @htuch @lambdai

**Review by lizan — APPROVED** [review]

**Review by lambdai — DISMISSED** [review]
Thanks! I find the critical listener update in the integration test.

File: test/integration/listener_lds_integration_test.cc [hunk 1]
```diff
@@ -257,5 +258,76 @@ TEST_P(ListenerIntegrationTest, BasicSuccess) {
...
+    sendRdsResponse(fmt::format(route_config_tmpl, route_table_name_, "cluster_0"),
+                    absl::StrCat(version));
+
+    test_server_->waitForCounterGe("listener_manager.listener_create_success", version);
+
+    // Wait for the client to be disconnected.
+    ASSERT_TRUE(codec_client_->waitForDisconnect());
+    // Make a new connection to the new listener.
+    codec_client_ = makeHttpConnection(lookupPort(listener_name_));
+    int response_size = 800;
```
Comment (htuch) [inline test/integration/listener_lds_integration_test.cc:318]: Nit: prefer `const uint32_t` etc. for Envoy style.

**Review by htuch — COMMENTED** [review]
LGTM, thanks!

File: test/integration/listener_lds_integration_test.cc (hunk 1 above)
Comment (penguingao) [reply test/integration/listener_lds_integration_test.cc]: Done

**Review by ggreenway — APPROVED** [review]

**Pull request merged**